ceno_host/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+    items.iter()
+}
+
+pub fn read<'a, T>(i: &mut CenoIter<'a>) -> &'a T
+where
+    T: Portable + for<'b> CheckBytes<HighValidator<'b, Error>>,
+{
+    let bytes = i.next().unwrap();
+    rkyv::access::<T, Error>(bytes).unwrap()
+}
+
+pub fn read_slice<'a>(i: &mut CenoIter<'a>) -> &'a [u8] {
+    let bytes = i.next().unwrap();
+    bytes.as_slice()
+}
+
+#[derive(Archive, Deserialize, Serialize, Debug, PartialEq)]
+#[rkyv(
+    // This will generate a PartialEq impl between our unarchived
+    // and archived types
+    compare(PartialEq),
+    // Derives can be passed through to the generated type:
+    derive(Debug),
+)]
+struct Test {
+    int: u32,
+    string: String,
+    option: Option<Vec<i32>>,
+}
+
+#[derive(Archive, Deserialize, Serialize, Debug, PartialEq)]
+#[rkyv(
+    // This will generate a PartialEq impl between our unarchived
+    // and archived types
+    compare(PartialEq),
+    // Derives can be passed through to the generated type:
+    derive(Debug),
+)]
+struct Toast {
+    stuff: Option<Vec<String>>,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::Rng;
+    use rkyv::{
+        deserialize,
+        rancor::{Error, Failure},
+        to_bytes,
+        util::AlignedVec,
+    };
+
+    /// The equivalent of this function would run in the host.
+    ///
+    /// We create three different items, and show that we can read them back in `consume`.
+    pub fn make_stdin() -> AlignedVec {
+        let mut stdin = CenoStdin::default();
+        stdin
+            .write(&Test {
+                int: 0xDEAD_BEEF,
+                string: "hello world".to_string(),
+                option: Some(vec![1, 2, 3, 4]),
+            })
+            .unwrap();
+        stdin.write(&0xaf_u8).unwrap();
+        stdin
+            .write(&Toast {
+                stuff: Some(vec!["hello scroll".to_string()]),
+            })
+            .unwrap();
+        stdin.finalise().unwrap()
+    }
+
+    /// The equivalent of this function would run in the guest.
+    ///
+    /// `buf` would be the memory mapped region for private hints.
+    pub fn consume(buf: AlignedVec) {
+        let archived = rkyv::access::<ArchivedCenoStdin, Failure>(&buf[..]).unwrap();
+        // This iterator would live in a mutable static behind the scenes,
+        // so that we don't have to pass it around.
+        let mut iter = prepare(archived);
+
+        let test1 = read::<ArchivedTest>(&mut iter);
+        assert_eq!(test1, &Test {
+            int: 0xDEAD_BEEF,
+            string: "hello world".to_string(),
+            option: Some(vec![1, 2, 3, 4]),
+        });
+        let number = read::<u8>(&mut iter);
+        assert_eq!(number, &0xaf_u8);
+        let test2 = read::<ArchivedToast>(&mut iter);
+        assert_eq!(test2, &Toast {
+            stuff: Some(vec!["hello scroll".to_string()]),
+        });
+    }
+
+    #[test]
+    fn test_prepare_and_consume_items() {
+        let stdin = make_stdin();
```

Comment:
IIUC line 139 & 140 is the host/guest boundary.
In real case, we will write `stdin` to a file then load in guest memory.

Then further question, why left/right padding matters?
In guest we load `stdin` into memory region with fixed start & dynamic end (end can be different per execution trace). Then in guest we just need get start/end, then restore AlignedVec from this specific address `mem[start:end]`. In this case, the padding is transparent & agnostic to `rkyv`?

Reply:
Introduce left padding is not free in terms of complexity for our current proving system, especially the verifier logic. It make the verifier memory range overlap check harder and more error pruning. I would suggest we better encaptulate it in SDK, as it's just minor effort, and I can already foresee it simplicity.

We can have end address defined in first word, then retrieved as `let end_hint_addr = mem[HINT_IO_REGION_START]`, then use mem[HINT_IO_REGION_START + 4: end_hint_addr] to restore AlignedVec. With that, there is almost no sacrificing with efficient, but we keep backend simple and consistent.

Reply:
I have a prototype of the new version on GitHub now.  It's a bit more of a mess, because instead of letting rkyv handle two layers of serialisation, we handle the outer layer ourselves, so that we can put its 'entry point' at the start of the hints memory region.

Reply:
You can already have a look at the new version, if you are feeling brave, but at the moment it's held together with hope and duct-tape.  Let me apply some polish to make it more palatable.

Reply:
It's saner now, and you can have a look.